[clangd] Implement simple folding of preprocessor branches (rebased)

sr-tream opened the pull request:
It's a rebased PR #80592 without pseudo-parser usage.

Extract directive branches information from DirectiveTree, fold branches that don't end with eof.

Fixes https://github.com/clangd/clangd/issues/1661

HighCommander4 commented on clang-tools-extra/clangd/support/DirectiveTree.h:
```diff
@@ -124,6 +124,9 @@ llvm::raw_ostream &operator<<(llvm::raw_ostream &,
 /// The choices are stored in Conditional::Taken nodes.
 void chooseConditionalBranches(DirectiveTree &, const TokenStream &Code);
 
+std::vector<Token::Range> pairDirectiveRanges(const DirectiveTree &Tree,
```
Please add a documentation comment similar to other function declarations in this header

HighCommander4 reviewed (changes requested):
This is my first time looking at this code (SemanticSelection.cpp / DirectiveTree.cpp) so I'm not sure if I'm the best reviewer, but on a high level what this patch is doing seems reasonable to me.

Some unit tests for the change would be great. We have some existing unit tests for folding ranges in [SemanticSelectionTests.cpp](https://searchfox.org/llvm/rev/1bd5f34d76b788de48e95d2f9c1179cf2685fbe5/clang-tools-extra/clangd/unittests/SemanticSelectionTests.cpp) which you can extend.

144026 committed:
[clangd] Implement simple folding of preprocessor branches

Extract directive branches information from DirectiveTree, fold branches
that don't end with eof.

Fixes https://github.com/clangd/clangd/issues/1661

bricle commented:
hope you guys are doing well, really need this feature, thank you for your contribution!

HighCommander4 commented:
Superseded by https://github.com/llvm/llvm-project/pull/140959.
